src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -579,5 +581,60 @@ public void ChainHasOneCheckPointAndAssumeValid_ChainsWithCheckpointButMissedAss
             result.DownloadFrom.HashBlock.Should().Be(listOfCurrentChainHeaders.Skip(2).First().GetHash());
             result.DownloadTo.HashBlock.Should().Be(listOfCurrentChainHeaders.Last().GetHash());
         }
+
+        /// <summary>
+        /// Issue 17 @ We advanced consensus tip (CT) and there are some partially validated (PV) headers after the CT.
+        /// Now we receive headers that are after the last PV header and include assume valid (AV). Make sure that those
+        /// headers that are before the AV header and after the last PV are all marked as AV. 
+        /// </summary>
+        [Fact]
+        public void ChainHasPartiallyValidatedAfterConsensusTip_NewHeadersWithAssumeValidPresented_CorrectHeadersAreMarkedAsAssumedValid()
+        {
+            // Chain header tree setup.
+            // Initial chain has 4 header with the consensus tip at header 4.
+            // Example: fv1=fv2=fv3=fv4 (fv - fully validated).
+            const int initialChainSize = 4;
+            TestContext ctx = new TestContextBuilder().WithInitialChain(initialChainSize).Build();
+            ChainedHeaderTree cht = ctx.ChainedHeaderTree;
+            ChainedHeader initialChainTip = ctx.InitialChainTip;
+
+            // Extend the chain with 2 partially validated headers.
+            // Example: fv1=fv2=fv3=(fv4)=pv5=pv6 (pv - partially validated).
+            const int partiallyValidatedHeadersCount = 2;
+            initialChainTip = ctx.ExtendAChain(partiallyValidatedHeadersCount, initialChainTip);
+            
+            // Chain is presented by peer 1.
+            // Mark pv5 and pv6 as partially validated.
+            List<BlockHeader> listOfCurrentChainHeaders =
+                ctx.ChainedHeaderToList(initialChainTip, partiallyValidatedHeadersCount);
+            ConnectNewHeadersResult result = cht.ConnectNewHeaders(1, listOfCurrentChainHeaders);
+            initialChainTip = result.Consumed;
+            initialChainTip.BlockValidationState = ValidationState.PartiallyValidated;
+            initialChainTip.Previous.BlockValidationState = ValidationState.PartiallyValidated;
+
+            // Extend the chain with 6 normal headers, where header at the height 9 is an "assumed valid" header.
+            // Example: fv1=fv2=fv3=(fv4)=pv5=pv6=h7=h8=(av9)=h10=h11=h12 (av - assumed valid).
+            const int extensionHeadersCount = 6;
+            initialChainTip = ctx.ExtendAChain(extensionHeadersCount, initialChainTip);
+            ChainedHeader assumedValidHeader = initialChainTip.GetAncestor(9);
+            assumedValidHeader.BlockValidationState = ValidationState.AssumedValid; 
```

Comment:
You should not set this. This is part of the job that is supposed to be done in connect new headers and which should be asserted.

Reply:
If I don’t set it, it doesn’t seem to work.

Reply:
Then it is a bug, but definitely not set it manually.

Reply:
It would be very unfortunate if we created this test or any other in a way that they passes with incorrect implementation of what is being tested.

Reply:
I agree, but usually the person writing the code writes a test as he/she knows what he is writing. In fact, tests should be written first. I am retrofitting a unit test to an existing code with 60-70% of understanding of what it is supposed to do and unfortunately I don’t have much experience with fullnode codebase. Eventually, I will get there.